buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java

```diff
@@ -470,17 +470,17 @@ private void consolidateIfNeeded() {
     }
 
     private void checkComponentIndex(int cIndex) {
-        ensureAccessible();
         if (cIndex < 0 || cIndex > componentCount) {
+            ensureAccessible();
```

Comment:
by moving it into the if aren't we changing the behaviour? and why check it only to throw an IOOBE?

Reply:
Yeah I think we should change this back @njhill

Reply:
@johnou @normanmaurer this was done in conjunction with setting `componentCount` to -1 upon deallocation - meaning the existing component index check catches both failure cases and the `ensureAccessible()` inside will ensure IRCE is thrown instead of IIOBE when appropriate.

However, benchmarking of this now that other changes have been made shows it provides negligible if any benefit, so I have reverted it in the latest commit (and just make use of the existing boolean `freed` field).